enso/experiment/rationalized.py

```diff
@@ -306,3 +313,199 @@ def _featurize(self, doc):
+                    doc_tensor[i, :doc_vector.shape[0], :] = doc_vector
+                    rationale_tensor[i, :rationale_target.shape[0], :] = rationale_target
+                
+                length_mask = np.zeros((bsize, max_length))
+                for i in range(bsize):
+                    length_mask[i, :lengths[i]] = 1.
+
+                yield (doc_tensor, rationale_tensor, target_slice, length_mask)
+
+    def predict_iter(self, doc_vectors, batch_size=100):
+        n_total = len(doc_vectors)
+        for batch_start in range(0, n_total, batch_size):
+            batch_end = min(batch_start + batch_size, n_total)
+            bsize = batch_end - batch_start
+            doc_vector_slice = doc_vectors[batch_start:batch_end]
+            lengths = list(map(len, doc_vector_slice))
+            max_length = max(lengths)
+            doc_tensor = np.zeros((bsize, max_length, 300))
+            for i, doc_vector in enumerate(doc_vector_slice):
+                doc_tensor[i, :doc_vector.shape[0], :] = doc_vector
+            
+            length_mask = np.zeros((bsize, max_length))
+            for i in range(bsize):
+                length_mask[i, :lengths[i]] = 1.
+
+            yield (doc_tensor, length_mask)
+
+    def featurize_x(self, X):
+        doc_vectors = []
+        for text in X:
+            doc = self.NLP(text, disable=['ner', 'tagger', 'textcat'])
+            word_vectors = np.zeros((len(doc), 300))
+            for i, token in enumerate(doc):
+                word_vectors[i,:] = token.vector
+            doc_vectors.append(word_vectors)
+        return doc_vectors
+
+    def featurize_x_y(self, X, Y):
+        doc_vectors = []
+        rationale_targets = []
+        targets = []
+        for text, (rationales, class_target) in zip(X, Y):
+            doc = self.NLP(text, disable=['ner', 'tagger', 'textcat'])
+            word_vectors = np.zeros((len(doc), 300))
+            rationale_one_hot = np.zeros((len(doc), self.n_classes))
+            target_one_hot = self.target_encoder.transform([class_target])[0]
+            for i, token in enumerate(doc):
+                word_vectors[i,:] = token.vector
+                if self._token_in_rationales(token, rationales):
+                    rationale_one_hot[i,:] = target_one_hot
+
+            doc_vectors.append(word_vectors)
+            rationale_targets.append(rationale_one_hot)
+            targets.append(target_one_hot)
+
+        # (batch, seq, hidden) * (batch, seq, one_hot)
+        stacked_docs = np.vstack(doc_vectors)
+        stacked_targets = np.vstack(rationale_targets)
+
+        rationale_proto = np.einsum('sh,st->ht', stacked_docs, stacked_targets)
+        rationale_proto /= np.einsum('st->t', stacked_targets)
+
+        return doc_vectors, rationale_targets, targets, rationale_proto
+        
+      
+    def fit(self, X, Y):
+        # No rationale means everything is a rationale
```

Comment:
Need to CV over batch_size, n_epochs, probably an L1/L2 weight penalty.  Maybe need early stopping or similar but not sure because it would probably have to be on train set (not enough data to split out to val).